Posts/2021/09/my-crescendo-journey.md

```diff
@@ -0,0 +1,158 @@
+Crescendo module and maybe I should just learn it and do the presentation myself.
+
+## What is Crescendo?
+
+Crescendo is an experimental module developed by [Jim Truher][jim], one of the main developers of
+PowerShell. Crescendo provides a framework to rapidly develop PowerShell cmdlets to wrap native
+commands, regardless of platform. The goal of a Crescendo-based module is to create PowerShell
+cmdlets that use a native command-line tool, but unlike the tool, return PowerShell objects instead
+of plain text.
+
+## How I got started
+
+When I first heard about Crescendo, I thought:
+
+> _So what. I've written wrapper modules like this before. How is this going to help me?_
+
+But I knew there must be more to it for Jim to invest this much time and effort into it, and I
+wanted something to present at the user group meeting.
+
+So, I started by reading the blog posts about Crescendo and looking a some examples in the
+[repository][repo].
+
+### How Crescendo works
+
+The to create a module using the Crescendo framework you have to create two main components:
+
+- A JSON configuration file that describes the cmdlets you want
+- Output handler functions that parse the output from the native command and return objects
+
+Initially, the parsing code had to be embedded in the JSON file, which made writing and formatting
+the code very difficult. But, in the [Preview 3][blog3] release, Jim added to the ability to create
+your output handler code in a function or a script file, making it much easier to manage.
+
+Alright! Writing the PowerShell functions is something I am more comfortable with, so that was my
+next step.
+
+## Writing the parser functions
+
+To create the parser functions I had to know what the output looked like for all of the possible
+command combinations of `vssadmin.exe`. I looked at the help provided by `vssadmin` and captured the
+output for each subcommand in a separate file. I used these output files to design and implement a
+parsing function for each subcommand.
+
+Now, on to the configuration file.
+
+## Creating the JSON configuration
+
+For this I used the example from the [blog post][blog3] as a template. I also looked at the
+`Get-InstalledPackage` example from the [Preview 2][blog2] blog post to see how the native commands
+were referenced. For my first cmdlet I started with this JSON configuration:
+
+```json
+{
+    "$schema": "./Microsoft.PowerShell.Crescendo.Schema.json",
+    "Commands": [
+         {
+            "Verb": "Get",
+            "Noun": "VssProvider",
+            "OriginalName": "$env:Windir/system32/vssadmin.exe",
+            "OriginalCommandElements": [
+                "list",
+                "providers"
+            ],
+            "OutputHandlers": [
+                {
+                    "ParameterSetName": "Default",
+                    "HandlerType": "Function",
+                    "Handler": "ParseProvider"
+                }
+            ],
+         }
+    ]
+}
+```
+
```

Comment:
you might add your example with the multiple parameter sets and your clever support of `/MaxSize=Unbounded`

Reply:
Yeah, I am not sure I want to go there in this blog post. I may have a follow where I examine the configuration and code I wrote.